client/core/exchangeratefetcher.go

```diff
@@ -0,0 +1,270 @@
+// also available online at https://blueoakcouncil.org/license/1.0.0.
+
+package core
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strings"
+	"sync"
+	"time"
+
+	"decred.org/dcrdex/dex"
+)
+
+const (
+	// DefaultFiatCurrency is the currency for displaying assets fiat value.
+	DefaultFiatCurrency = "USD"
+	// fiatRateRequestInterval is the amount of time between calls to the exchange API.
+	fiatRateRequestInterval = 5 * time.Minute
+	// fiatRateDataExpiry : Any data older than fiatRateDataExpiry will be discarded.
+	fiatRateDataExpiry = 60 * time.Minute
+
+	// Tokens. Used to identify fiat rate source.
+	messari       = "Messari"
+	coinpaprika   = "Coinpaprika"
+	dcrdataDotOrg = "dcrdata"
+)
+
+var (
+	dcrDataURL     = "https://explorer.dcrdata.org/api/exchangerate"
+	coinpaprikaURL = "https://api.coinpaprika.com/v1/tickers/%s"
+	messariURL     = "https://data.messari.io/api/v1/assets/%s/metrics/market-data"
+	btcBipId, _    = dex.BipSymbolID("btc")
+	dcrBipId, _    = dex.BipSymbolID("dcr")
+)
+
+// fiatRateFetchers is the list of all supported fiat rate fetchers.
+var fiatRateFetchers = map[string]rateFetcher{
+	coinpaprika:   fetchCoinpaprikaRates,
+	dcrdataDotOrg: fetchDcrdataRates,
+	messari:       fetchMessariRates,
+}
+
+// fiatRateInfo holds the fiat rate and the last update time for an
+// asset.
+type fiatRateInfo struct {
+	rate       float64
+	lastUpdate time.Time
+}
+
+// rateFetcher can fetch fiat rates for assets from an API.
+type rateFetcher func(context context.Context, logger dex.Logger, assets map[uint32]*SupportedAsset) map[uint32]float64
+
+type commonRateSource struct {
+	mtx         sync.RWMutex
+	lastRequest time.Time
+	log         dex.Logger
+	fetchRates  rateFetcher
+	fiatRates   map[uint32]*fiatRateInfo
+}
+
+// logRequest sets the lastRequest time.Time.
+func (source *commonRateSource) logRequest() {
+	source.mtx.Lock()
+	defer source.mtx.Unlock()
+	source.lastRequest = time.Now()
+}
+
+// lastTry is the last time the fiat rate fetcher made a request.
+func (source *commonRateSource) lastTry() time.Time {
+	source.mtx.RLock()
+	defer source.mtx.RUnlock()
+	return source.lastRequest
+}
+
+// isExpired checks the last update time for all fiat rates against the
+// provided expiryTime.
+func (source *commonRateSource) isExpired(expiryTime time.Duration) bool {
+	source.mtx.Lock()
+	defer source.mtx.Unlock()
+	var expiredCount int
+	for _, rateInfo := range source.fiatRates {
+		if time.Since(rateInfo.lastUpdate) > expiryTime {
+			expiredCount++
+		}
+	}
+	totalFiatRate := len(source.fiatRates)
+	return expiredCount == totalFiatRate && totalFiatRate != 0
+}
+
+// assetRate returns the fiat rate information for the assetID specified.
+func (source *commonRateSource) assetRate(assetID uint32) *fiatRateInfo {
+	source.mtx.Lock()
+	defer source.mtx.Unlock()
+	rateInfo := source.fiatRates[assetID]
+	var fiatRateInfo fiatRateInfo
+	if rateInfo != nil {
+		fiatRateInfo = *rateInfo
+	}
+	return &fiatRateInfo
```

Comment:
I think this will no longer be protected by the lock if the caller gets the pointer to the struct in the map.

> it won't be modified, will it?

Maybe.

Reply:
> I think this will no longer be protected by the lock if the caller gets the pointer to the struct in the map.

What does the mutex aim to protect?  The map, the elements in the map, or both?  This is an extremely important distinction.  It should be documented clearly.  Being unclear about this is a very common source of bugs.

> Maybe.

Be clear and specific please.  It should be known and documented.  I see the map entry being replcaed entirely in `refreshRates`, not modified.

Reply:
I think this was my suggestion. The returned value could be altered changing the value in the map if we just return the pointer. Currently the value is only read so this is not necessary... now. I think either doing it this way or leaving a note in the comments that the value should not be altered is good to prevent possible bugs in the future.

Apologies if this is overthinking it.

Reply:
> What does the mutex aim to protect? 

The map, in `refreshRates` the`fiatRatesInfo` for an asset  in `source.fiatRates` is replaced if a new rate was returned.

Reply:
We don't need the copy. You would only need a copy if you expected the `fiatRateInfo` fields to be modified, but once created, the `fiatRateInfo` fields are read-only. You're only protecting the map. IMO, we don't even need to document that. It's sort of the default. But if other's want a doc, you should.

Reply:
Sorry @chappjc. I wasn't calling out your request for docs as unusual or off-base. On the contrary, I think as soon as there appears to be some confusion amongst developers around any topic, more docs are righteous.

Reply:
made changes to `source.assetRate` and dropeda comment.